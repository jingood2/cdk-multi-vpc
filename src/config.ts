//import * as chalk from 'chalk';

export const envVars = {
  COMPANY_NAME: 'skcnc',
  PROJECT_NAME: 'network',
  SOURCE_PROVIDER: 'GITHUB',
  GITHUB_TOKEN: 'atcl/jingood2/github-token',
  REPO_OWNER: process.env.REPO_OWNER || 'jingood2',
  REPO: process.env.REPO_NAME || 'cdk-vpc-pipeline',
  ENABLE_TGW_USE: 'true',
  ENABLE_VPN_USE: 'false',
  ENABLE_ZCP_USE: 'true',
  DMZ_VPC_CIDR_BLOCK: '10.0.0.0/18',
  DEV_VPC_CIDR_BLOCK: '10.1.0.0/18',
  PROD_VPC_CIDR_BLOCK: '10.2.0.0/18',
  SHARED_VPC_CIDR_BLOCK: '192.168.0.0/18',
  ONPREMISE_CIDR_BOCK: '192.168.0.0/18',
  VPC_ENV_INFO: [
    { DMZ: { VPC_CIDR_BLOCK: '10.0.0.0/18', TGW_ROUTE: ['10.1.0.0/18'] } },
    { DEV: { VPC_CIDR_BLOCK: '10.1.0.0/18', TGW_ROUTE: ['0.0.0.0/0'] } },
    { SHARED: { VPC_CIDR_BLOCK: '192.168.0.0/18', TGW_ROUTE: ['0.0.0.0/0'] } },
  ],
  CGW_PUBLIC_IP: '11.23.22.0/32',
  BRANCH: 'develop',
};

export function validateEnvVariables() {
  for (let variable in envVars) {
    if (!envVars[variable as keyof typeof envVars]) {
      throw Error(
        //chalk.red(`[app]: Environment variable ${variable} is not defined!`),
        //chalk.chalkStderr(`[app]: Environment variable ${variable} is not defined!`),
        `[app]: Environment variable ${variable} is not defined!`,
      );
    }
  }
}