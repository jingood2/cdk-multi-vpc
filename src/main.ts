import { App, Stack, StackProps } from 'aws-cdk-lib';
//import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';
import { envVars, validateEnvVariables } from './config';
import { TgwConstruct } from './lib/tgw-construct';
import { VpcConstruct } from './lib/vpc-construct';
import { VpnConstruct } from './lib/vpn-construct';

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);


    const tgwStack = new TgwConstruct(this, 'TgwConstruct');

    // define resources here...
    // 1. VPC
    envVars.VPC_ENV_INFO.forEach((vpc) => {
      new VpcConstruct(this, `${vpc.STAGE}-VPC`, {
        //env: props.env,
        //env: { account: '037729278610', region: 'ap-northeast-2' },
        customProps: {
          company: `${envVars.COMPANY_NAME}`,
          project: `${envVars.PROJECT_NAME}`,
          stage: vpc.STAGE,
          tgwCidrBlock: vpc.TGW_ROUTE,
          useNATInstance: true,
        },
        vpcProps: {
          cidr: vpc.VPC_CIDR_BLOCK,
          enableDnsHostnames: true,
          enableDnsSupport: true,
          natGateways: vpc.NGW,
          maxAzs: 2,
          subnetConfiguration: vpc.SUBNET_CONFIG,
        },
      }).addDependency(tgwStack);

    });
    /* new VpcConstruct(this, 'DmzVPC', {
      //env: props.env,
      //env: { account: '037729278610', region: 'ap-northeast-2' },
      customProps: {
        company: `${envVars.COMPANY_NAME}`,
        project: `${envVars.PROJECT_NAME}`,
        stage: 'dmz',
        tgwCidrBlock: envVars.VPC_ENV_INFO.DMZ.TGW_ROUTE,
        useNATInstance: true,
      },
      vpcProps: {
        cidr: envVars.VPC_ENV_INFO.DMZ.VPC_CIDR_BLOCK,
        enableDnsHostnames: true,
        enableDnsSupport: true,
        natGateways: 1,
        maxAzs: 2,
        subnetConfiguration: [
          { name: 'pub', subnetType: ec2.SubnetType.PUBLIC, cidrMask: 28 },
          { name: 'pri', subnetType: ec2.SubnetType.PRIVATE_WITH_NAT, cidrMask: 24 },
        ],
      },
    }).addDependency(tgwStack);

    // 1. VPC
    new VpcConstruct(this, 'DevVpc', {
      //env: props.env,
      //env: { account: '037729278610', region: 'ap-northeast-2' },
      customProps: {
        company: `${envVars.COMPANY_NAME}`,
        project: `${envVars.PROJECT_NAME}`,
        stage: 'dev',
        tgwCidrBlock: ['0.0.0.0/0'],
        useNATInstance: false,
      },
      vpcProps: {
        cidr: envVars.DEV_VPC_CIDR_BLOCK,
        enableDnsHostnames: true,
        enableDnsSupport: true,
        natGateways: 0,
        maxAzs: 2,
        subnetConfiguration: [
          { name: 'pri', subnetType: ec2.SubnetType.PRIVATE_ISOLATED, cidrMask: 22 },
        ],
      },
    }).addDependency(tgwStack);

    // 1. VPC
    new VpcConstruct(this, 'SharedVpc', {
      //env: props.env,
      //env: { account: '037729278610', region: 'ap-northeast-2' },
      customProps: {
        company: `${envVars.COMPANY_NAME}`,
        project: `${envVars.PROJECT_NAME}`,
        stage: 'shared',
        tgwCidrBlock: ['0.0.0.0/0'],
        useNATInstance: false,
      },
      vpcProps: {
        cidr: envVars.SHARED_VPC_CIDR_BLOCK,
        enableDnsHostnames: true,
        enableDnsSupport: true,
        natGateways: 0,
        maxAzs: 2,
        subnetConfiguration: [
          { name: 'pri', subnetType: ec2.SubnetType.PRIVATE_ISOLATED, cidrMask: 22 },
        ],
      },
    }).addDependency(tgwStack); */

    if (envVars.ENABLE_VPN_USE == 'true') {
      new VpnConstruct(this, 'OnPremA', {
      // openswan 설치된 ec2의 eip 입력
        vpnIpAddress: envVars.CGW_PUBLIC_IP,
        bgpArn: 65001,
        onPremiseCidrBlock: envVars.ONPREMISE_CIDR_BOCK,
      }).addDependency(tgwStack);
    }

  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

validateEnvVariables();
new MyStack(app, 'multivpc-stack', { env: devEnv });
// new MyStack(app, 'my-stack-prod', { env: prodEnv });

app.synth();