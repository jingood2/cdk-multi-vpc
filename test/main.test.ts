import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MyStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();

  const devEnv = {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  };
  const stack = new MyStack(app, 'test', { env: devEnv });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});