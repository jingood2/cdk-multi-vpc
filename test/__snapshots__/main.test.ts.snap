// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DevVpcNestedStackDevVpcNestedStackResource7478DE5D": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "TgwConstructNestedStackTgwConstructNestedStackResourceBDDD946A",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cz-ext1",
            "Value": " ",
          },
          Object {
            "Key": "cz-ext2",
            "Value": " ",
          },
          Object {
            "Key": "cz-ext3",
            "Value": " ",
          },
          Object {
            "Key": "cz-org",
            "Value": "skcnc",
          },
          Object {
            "Key": "cz-project",
            "Value": "network",
          },
          Object {
            "Key": "cz-stage",
            "Value": "dev",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.ap-northeast-2.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/cdk-hnb659fds-assets-037729278610-ap-northeast-2/f92c1442c6d1ea62635daee6490d19bb5a41c55a0580f86eb9a4db0b96632dfa.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "DmzVPCNestedStackDmzVPCNestedStackResource6D61C35C": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "TgwConstructNestedStackTgwConstructNestedStackResourceBDDD946A",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cz-ext1",
            "Value": " ",
          },
          Object {
            "Key": "cz-ext2",
            "Value": " ",
          },
          Object {
            "Key": "cz-ext3",
            "Value": " ",
          },
          Object {
            "Key": "cz-org",
            "Value": "skcnc",
          },
          Object {
            "Key": "cz-project",
            "Value": "network",
          },
          Object {
            "Key": "cz-stage",
            "Value": "dmz",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.ap-northeast-2.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/cdk-hnb659fds-assets-037729278610-ap-northeast-2/25331edf212a1573e0bbac7b90d4ac112fdb5dca2f09853edbc608b04969c600.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "SharedVpcNestedStackSharedVpcNestedStackResource28D5D41B": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "TgwConstructNestedStackTgwConstructNestedStackResourceBDDD946A",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cz-ext1",
            "Value": " ",
          },
          Object {
            "Key": "cz-ext2",
            "Value": " ",
          },
          Object {
            "Key": "cz-ext3",
            "Value": " ",
          },
          Object {
            "Key": "cz-org",
            "Value": "skcnc",
          },
          Object {
            "Key": "cz-project",
            "Value": "network",
          },
          Object {
            "Key": "cz-stage",
            "Value": "shared",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.ap-northeast-2.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/cdk-hnb659fds-assets-037729278610-ap-northeast-2/626dc0a57ce1c786e0f8061b920b2a02e08e21b4f16becdce3cbdbd03a2ae584.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "TgwConstructNestedStackTgwConstructNestedStackResourceBDDD946A": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.ap-northeast-2.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/cdk-hnb659fds-assets-037729278610-ap-northeast-2/072d985dbe7c88448fca3f4b2ef4dc255a54b7ab7cdde31f2ef20a4b918fbb18.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
